### PR TITLE
Add test accessor + more verification on tests and fix flakiness

### PR DESCRIPTION
**Pull request by dibarbet**
Followup for https://github.com/dotnet/roslyn/pull/42341

Add test accessor + verify
Fix flakiness by ordering stream json rpc messages

### CODE REVIEW

File: src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs [hunk 1]
```diff
@@ -331,29 +343,37 @@ public async Task ClearAllDiagnosticsForMappedFileToManyDocumentsTestAsync()
...
-                {
-                    await languageServer.PublishDiagnosticsAsync(document).ConfigureAwait(false);
-                }
+            using var jsonRpc = new JsonRpc(clientStream, clientStream, callback);
 
-                await callback.CallbackCompletedTask.Task.ConfigureAwait(false);
+            // The json rpc messages won't necessarily come back in order by default.
+            // So use a synchronization context to preserve the original ordering.
+            // https://github.com/microsoft/vs-streamjsonrpc/blob/bc970c61b90db5db135a1b3d1c72ef355c2112af/doc/resiliency.md#when-message-order-is-important
+            jsonRpc.SynchronizationContext = new RpcOrderPreservingSynchronizationContext(); ;
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Test.Next/Services/LspDiagnosticsTests.cs:363]:
```suggestion
            jsonRpc.SynchronizationContext = new RpcOrderPreservingSynchronizationContext();
```

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs [hunk 2]
```diff
@@ -390,5 +393,43 @@ private bool IncludeDiagnostic(DiagnosticData diagnostic)
...
+
+            internal ImmutableArray<Uri> GetFileUrisInPublishDiagnostics()
+                => _server._publishedFileToDiagnostics.Keys.ToImmutableArray();
+
+            internal ImmutableArray<DocumentId> GetDocumentIdsInPublishedUris()
+                => _server._documentsToPublishedUris.Keys.ToImmutableArray();
+
+            internal IImmutableSet<Uri> GetFileUrisForDocument(DocumentId documentId)
+            {
+                if (_server._documentsToPublishedUris.TryGetValue(documentId, out var uris))
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:416]: Didn't you add a Didn't you add a GetValueOrDefault?

Comment (dibarbet) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:416]: 🤦‍♂

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs [hunk 3]
```diff
@@ -390,5 +393,36 @@ private bool IncludeDiagnostic(DiagnosticData diagnostic)
             var linePositionSpan = DiagnosticData.GetLinePositionSpan(diagnosticDataLocation, text, useMapped: true);
             return ProtocolConversions.LinePositionToRange(linePositionSpan);
         }
+
+        internal TestAccessor GetTestAccessor() => new TestAccessor(this);
+
+        internal class TestAccessor
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs:399]: 📝 Test accessors should always be a `readonly struct`

**Pull request merged**